File: 프로그래머스/1/172928. 공원 산책/공원 산책.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<string> park, vector<string> routes) {
    vector<int> answer;
    int x, y, num;
    x = -1; y = -1;
    for(int i = 0; i < park.size(); i++){
        for(int j = 0; j < park[i].size(); j++){
            if(park[i][j] == 'S'){
                x = i; y = j;
                break;
            }
        }
        if(x != -1) break;
    }
    
    int temp = 0;
    for(int i = 0; i < routes.size(); i++){
        temp = routes[i][2] - '0';
        if(routes[i][0] == 'E'){
            if(y + temp > park[0].size()-1) continue;
            else{
                num = y;
                for(int j = 0; j < temp; j++){
                    y++;
                    if(park[x][y] == 'X'){
                        y = num;
                        break;
                    }
                }
            }
        }
        else if(routes[i][0] == 'W'){
            if(y - temp < 0) continue;
            else{
                num = y;
                for(int j = 0; j < temp; j++){
                    y--;
                    if(park[x][y] == 'X'){
                        y = num;
                        break;
                    }
                }
            }
        }
        else if(routes[i][0] == 'S'){
            if(x + temp > park.size()-1) continue;
            else{
                num = x;
                for(int j = 0; j < temp; j++){
                    x++;
                    if(park[x][y] == 'X'){
                        x = num;
                        break;
                    }
                }
            }
        }
        else if(routes[i][0] == 'N'){
            if(x - temp < 0) continue;
            else{
                num = x;
                for(int j = 0; j < temp; j++){
                    x--;
                    if(park[x][y] == 'X'){
                        x = num;
                        break;
                    }
                }
            }
        }
    }
    answer.push_back(x);
    answer.push_back(y);
    return answer;
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.cs
using System;
using System.Collections.Generic;

public class Solution
{
int _H = 0;
int _W = 0;

public int[] solution(string[] park, string[] routes)
{
    int[] answer = new int[] { };

    int[] location = new int[2];                    //  현재 위치
    List<string[]> route = new List<string[]>();    // 이동할 경로

    _H = park.Length - 1;                           // 세로 길이
    _W = park[0].Length - 1;                        // 가로 길이

    // 이동할 칸 수 세팅
    for (int i = 0; i < routes.Length; i++)
    {
        route.Add(new string[] { routes[i].Split(' ')[0], routes[i].Split(' ')[1] });
    }

    // 시작 위치 찾기
    for(int i = 0; i < park.Length; i++)
    {
        if (park[i].IndexOf('S') >= 0)
        {
            location[0] = i;
            location[1] = park[i].IndexOf('S');
            break;
        }
    }

    //  이동
    for(int i = 0; i < route.Count; i++)
    {
        location = Walk(park, location, route[i][0], int.Parse(route[i][1]));
    }

    return location;
}

public int[] Walk(string[] _Park, int[] _Location, string op, int n)
    {
        if (op.ToUpper() == "W")
        {
            if (_Location[1] - n < 0)
            {
                return _Location;
            }

            for (int i = 1; i <= n; i++)
            {
                if (_Park[_Location[0]][_Location[1] - i] == 'X')
                {
                    return _Location;
                }
            }

            _Location[1] = _Location[1] - n;

        }

        else if (op.ToUpper() == "N")
        {
            if (_Location[0] - n < 0)
            {
                return _Location;
            }

            for (int i = 1; i <= n; i++)
            {
                if (_Park[_Location[0] - i][_Location[1]] == 'X')
                {
                    return _Location;
                }
            }

            _Location[0] = _Location[0] - n;

        }
        else if (op.ToUpper() == "E")
        {
            if (_Location[1] + n >  _W)
            {
                return _Location;
            }

            for (int i = 1; i <= n; i++)
            {
                if (_Park[_Location[0]][_Location[1] + i] == 'X')
                {
                    return _Location;
                }
            }
            _Location[1] = _Location[1] + n;

        }
        else if (op.ToUpper() == "S")
        {
            if (_Location[0] + n > _H)
            {
                return _Location;
            }

            for (int i = 1; i <= n; i++)
            {
                if (_Park[_Location[0] + i][_Location[1]] == 'X')
                {
                    return _Location;
                }
            }
            _Location[0] = _Location[0] + n;

        }

        return _Location;
    }
}